pkg/egresspolicy/manager.go

```diff
@@ -0,0 +1,461 @@
+					logfields.ENPName: config.id.Name,
+				}).Error("Error deleting pod mapping for policy.")
+				return
+			}
+		}
+		delete(manager.policyPods, id)
+	}
+	delete(manager.podDataStore, id)
+}
+
+func (manager *Manager) getLocalPodsForNamespace(ns *slim_corev1.Namespace) []*slim_corev1.Pod {
+	var retPods []*slim_corev1.Pod
+
+	podStore := manager.storeGetter.GetStore("pod")
+	for _, podItem := range podStore.List() {
+		pod, ok := podItem.(*slim_corev1.Pod)
+		if !ok {
+			continue
+		}
+
+		if pod.Namespace != ns.Name {
+			continue
+		}
+		retPods = append(retPods, pod)
+	}
+
+	return retPods
+}
+
+// OnUpdateNamespace is the handler for namespace label updates.
+func (manager *Manager) OnUpdateNamespace(ns *slim_corev1.Namespace) {
+	manager.mutex.Lock()
+	defer manager.mutex.Unlock()
+	// Ns labels have been updated, need to re-evaluate all pods within ns.
+	pods := manager.getLocalPodsForNamespace(ns)
+	if len(pods) == 0 {
+		return
+	}
+	for _, pod := range pods {
+		manager.OnUpdatePodLocked(pod, true /* removeOld */, true /* upsertNew */)
+	}
+}
+
+func (manager *Manager) getPodMetadata(pod *slim_corev1.Pod, podIPs []string) (*podMetadata, error) {
+	podId := types.NamespacedName{
+		Name:      pod.GetName(),
+		Namespace: pod.GetNamespace(),
+	}
+
+	namespace, err := manager.getNamespaceForPod(pod)
+	if namespace == nil || err != nil {
+		data := &podMetadata{
+			ips:    podIPs,
+			labels: pod.GetLabels(),
+			id:     podId,
+		}
+
+		// Keep id to pod data mapping up to date
+		manager.podDataStore[podId] = data
+		return data, nil
+	}
+
+	data := &podMetadata{
+		ips:      podIPs,
+		labels:   pod.GetLabels(),
+		nsLabels: namespace.GetLabels(),
+		id:       podId,
+	}
+
+	// Keep id to pod data mapping up to date
+	manager.podDataStore[podId] = data
+	return data, nil
+}
+
+func (manager *Manager) getNamespaceForPod(pod *slim_corev1.Pod) (*slim_corev1.Namespace, error) {
+	nsName := pod.GetNamespace()
+	nsStore := manager.storeGetter.GetStore("namespace")
+	nsItem, exists, _ := nsStore.GetByKey(pod.GetNamespace())
+	if !exists {
+		log.WithFields(logrus.Fields{logfields.K8sNamespace: nsName}).Warn("Namespace does not exist.")
+		return nil, nil
+	}
+	return nsItem.(*slim_corev1.Namespace), nil
+}
+
+// isValidConfig validates the given policy config.
+func (manager *Manager) isValidConfig(config EgressPolicyConfig) error {
+	for _, policyConfig := range manager.policyConfigs {
+		if policyConfig.egressIP.String() == config.egressIP.String() {
+			return fmt.Errorf("CiliumEgressNatPolicy for egress IP %v already exists", config.egressIP.String())
+		}
+	}
+	return nil
+}
+
+// upsertPolicyPod updates or inserts to pod policy mapping for given policy config and pods,
+// it also upserts the pod data cache and egress map to keep in sync
+func (manager *Manager) upsertPolicyPod(config *EgressPolicyConfig, pods ...*podMetadata) error {
+	for _, pod := range pods {
+		podIp := net.ParseIP(pod.ips[0])
```

Comment:
a pod will and can have multiple IP addresses, is it something that we don't support? If we don't support it can we make it explicit to the user?

Reply:
Since the change to endpoints we do include updates for all the pod ips in the ipv4 field now.